core/src/main/java/org/elasticsearch/action/support/AutoCreateIndex.java

```diff
@@ -67,11 +69,13 @@ public boolean needToCheck() {
     /**
      * Should the index be auto created?
      */
-    public boolean shouldAutoCreate(String index, ClusterState state) {
+    public boolean shouldAutoCreate(String index, ClusterState state, IndexNameExpressionResolver resolver) {
         if (!needToCheck) {
             return false;
         }
-        if (state.metaData().hasConcreteIndex(index)) {
+        IndicesOptions options = IndicesOptions.fromOptions(true, true, true, true);
+        String[] concreteIndices = resolver.concreteIndices(state, options, index);
+        if (concreteIndices.length > 0) {
```

Comment:
Actually we already support index expression resolving in write apis. But it only did not if the index did not exist. This makes sure that index expressions work too in that case.

I think it is safe to support dynamic index expressions too in write apis, because if an expression resolve to more than one concrete index we fail. Also now with the date math index expressions it is especially useful (write into today's index).